src/renderers/dom/server/__tests__/ReactServerRendering-test.js

```diff
@@ -254,6 +254,71 @@ describe('ReactServerRendering', function() {
         'renderToString(): You must pass a valid ReactElement.'
       );
     });
+
+    it('should warn and ignore setState calls executed ' +
+        'after componentWillMount', function() {
+      spyOn(console, 'error');
+
+      var Component = React.createClass({
+        componentWillMount: function() {
+          setTimeout(() => {
+            this.setState({text: 'hello, world'});
+          });
+        },
+        render: function() {
+          return <div></div>;
+        },
+      });
+
+      ReactReconcileTransaction.prototype.perform = function() {
+        // We shouldn't ever be calling this on the server
+        throw new Error('Browser reconcile transaction should not be used');
+      };
+      ReactServerRendering.renderToString(
+        <Component />
+      );
+
+      jest.runAllTimers();
+      expect(setTimeout.mock.calls.length).toBe(1);
+      expect(console.error.calls.length).toBe(1);
+      expect(console.error.argsForCall[0][0]).toContain(
+        'setState(...): method calls executed after componentWillMount ' +
+        '(e.g. setTimeout() callbacks)  are ignored if component was ' +
+        'rendered on server.'
+      );
+    });
+
+    it('should warn and ignore all forceUpdate calls', function() {
+      spyOn(console, 'error');
+
+      var Component = React.createClass({
+        componentWillMount: function() {
+          this.forceUpdate();
+          setTimeout(() => {
+            this.forceUpdate();
+          });
+        },
+        render: function() {
+          return <div></div>;
+        },
+      });
+
+      ReactReconcileTransaction.prototype.perform = function() {
+        // We shouldn't ever be calling this on the server
+        throw new Error('Browser reconcile transaction should not be used');
+      };
+      ReactServerRendering.renderToString(
+        <Component />
+      );
+
+      var expectedError = 'forceUpdate(...): all method calls are ignored ' +
+        'if component was rendered on server.';
+      jest.runAllTimers();
+      expect(setTimeout.mock.calls.length).toBe(1);
+      expect(console.error.calls.length).toBe(2);
```

Comment:
We generally stick to warn-once where possible/practical, because otherwise a warning that fires often can drown out a warning that only fires infrequently (which results in people not seeing the less-frequent warning).

Let's add a global boolean instead of using _serverSideRendered, since the latter will be thrown away each render.